File: shared/types/src/annotation.ts
import type { SourceResult, CommentaryItem } from './source.js'

export type RiskLevel = 'high' | 'medium' | 'low'

export type AnnotationState = 'sourced' | 'limited' | 'unverified'

export interface TensionRating {
  /**
   * Weighted mean of (divergenceScore × relevanceScore) across all Tier 1/2 sources,
   * normalised to 0.0–1.0.
   * 0.0 = sources frame the claim consistently; 1.0 = sources diverge substantially.
   * Maps to a blue→red gradient in the UI. Never treated as a binary verdict.
   */
  score: number
  /** Number of Tier 1/2 sources the score is based on — shown as context ("based on N sources") */
  sourceCount: number
  /**
   * Accessibility label describing the score range without a verdict.
   * Shown on hover / for screen readers. Ranges:
   *   0.00–0.25 → "Sources largely frame this consistently"
   *   0.25–0.50 → "Sources show some variation in framing"
   *   0.50–0.75 → "Sources show notable framing differences"
   *   0.75–1.00 → "Sources show substantial framing differences"
   */
  label: string
}

export interface Claim {
  claimText: string
  charOffset: number
  charLength: number
  riskLevel: RiskLevel
  riskReasoning: string
  /** Optimised query for web search verification */
  searchQuery: string
}

export interface Annotation {
  claim: Claim
  state: AnnotationState
  /**
   * null when no Tier 1/2 sources were found, or when evaluationFailed is true.
   * If evaluationFailed is true, the user can retry via POST /v1/evaluate-claim.
   */
  tensionRating: TensionRating | null
  sources: SourceResult[]
  commentaryItems: CommentaryItem[]
  /** ISO timestamp of when this annotation was generated */
  generatedAt: string
  /**
   * True if source evaluation or framing analysis failed for this claim.
   * The UI shows a retry affordance; the claim can be re-run via POST /v1/evaluate-claim.
   * Other fields are still populated from claim extraction (claimText, riskLevel, etc.).
   */
  evaluationFailed?: boolean
}

File: services/annotator/internal/domain/annotation.go
package domain

import "fmt"

// TensionRating is a probabilistic assessment — never a verdict.
// Score is the weighted mean of (divergenceScore × relevanceScore) across all
// Tier 1/2 sources, normalised to 0.0–1.0. Maps to a blue→red gradient in the UI.
type TensionRating struct {
	// Score is 0.0 (sources align) to 1.0 (sources diverge substantially).
	Score float64
	// SourceCount is the number of Tier 1/2 sources the score is based on.
	SourceCount int
}

// Label returns an accessibility label describing the score range without a verdict.
// Language is always hedged and probabilistic — never "contradicts" or "false".
func (t TensionRating) Label() string {
	switch {
	case t.SourceCount == 0:
		return "No rated sources found"
	case t.Score < 0.25:
		return fmt.Sprintf("Sources largely frame this consistently (based on %d sources)", t.SourceCount)
	case t.Score < 0.50:
		return fmt.Sprintf("Sources show some variation in framing (based on %d sources)", t.SourceCount)
	case t.Score < 0.75:
		return fmt.Sprintf("Sources show notable framing differences (based on %d sources)", t.SourceCount)
	default:
		return fmt.Sprintf("Sources show substantial framing differences (based on %d sources)", t.SourceCount)
	}
}

// AnnotationState describes what Shirajitsu found for a given claim.
type AnnotationState string

const (
	StateSourced    AnnotationState = "sourced"
	StateLimited    AnnotationState = "limited"
	StateUnverified AnnotationState = "unverified"
)
